File: src/utils/storage/UStorage.ts
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { CloudDataStorage } from "./CloudDataStorage";
import { DefDataStorage } from "./DefDataStorage";
import { DataStorage } from "./declare/DataStorage";

//Типа UniversalStorage :D
export function UStorage<SchemaStorage extends object>(defaultValues: SchemaStorage, storageName = 'storage'): DataStorage<SchemaStorage> {
    type Keys = keyof typeof defaultValues;

    const proxy = getProxy();

    function getProxy() {
        //const isCloudSupported = html5 != undefined && Ads.bridge.storage.isSupported();
        const isCloudSupported = false;

        return isCloudSupported ? CloudDataStorage(defaultValues, storageName) : DefDataStorage(defaultValues, storageName);
    }

    load();

    function set<T extends Keys>(key: T, value: SchemaStorage[T]) {
        proxy.set(key, value);
    }

    function get<T extends Keys>(key: T): SchemaStorage[T] {
        return proxy.get(key);
    }

    function load() {
        proxy.load();
    }

    function clear() {
        proxy.clear();
    }

    return {
        set, get, load, clear, schemaObject: defaultValues
    };
}

File: app/utils/storage/UStorage.lua
local ____exports = {}
local ____CloudDataStorage = require("utils.storage.CloudDataStorage")
local CloudDataStorage = ____CloudDataStorage.CloudDataStorage
local ____DefDataStorage = require("utils.storage.DefDataStorage")
local DefDataStorage = ____DefDataStorage.DefDataStorage
function ____exports.UStorage(defaultValues, storageName)
    if storageName == nil then
        storageName = "storage"
    end
    local getProxy, load, proxy
    function getProxy()
        local isCloudSupported = false
        local ____isCloudSupported_0
        if isCloudSupported then
            ____isCloudSupported_0 = CloudDataStorage(defaultValues, storageName)
        else
            ____isCloudSupported_0 = DefDataStorage(defaultValues, storageName)
        end
        return ____isCloudSupported_0
    end
    function load()
        proxy.load()
    end
    proxy = getProxy()
    load()
    local function set(key, value)
        proxy.set(key, value)
    end
    local function get(key)
        return proxy.get(key)
    end
    local function clear()
        proxy.clear()
    end
    return {
        set = set,
        get = get,
        load = load,
        clear = clear,
        schemaObject = defaultValues
    }
end
return ____exports
